Simplify augmentations. (#4357)
dart-lang/language: working/augmentation-libraries/feature-specification.md

memberDeclarations ::= (metadata memberDeclaration)*

memberDeclaration ::= declaration
  | 'augment'? methodSignature functionBody

enumEntry ::= metadata 'augment'? identifier argumentPart?
  | metadata 'augment'? identifier typeArguments?
    '.' identifierOrNew arguments

declaration ::= 'external'? factoryConstructorSignature ';'
  | 'augment'? 'external' constantConstructorSignature ';'
  | 'augment'? 'external' constructorSignature ';'
  | 'augment'? 'external'? 'static'? getterSignature ';'
  | 'augment'? 'external'? 'static'? setterSignature ';'
  | 'augment'? 'external'? 'static'? functionSignature ';'
  | 'external' ('static'? finalVarOrType | 'covariant' varOrType) identifierList ';'
  | 'augment'? 'external'? operatorSignature ';'

>   | 'augment'? 'abstract' (finalVarOrType | 'covariant' varOrType) identifierList ';'

Author: > So no augmenting with a concrete variable declaration?

Yeah, my pitch is that we just don't support augmenting variable declarations or augmenting with variable declarations at all. If you have an abstract getter and you want to augment it with a backing store, you can always augment with a getter that returns the value of a differently named field.

> In which case, what's the point of augmenting a variable declaration at all, other than to add metadata. I guess that is a reason.

It's my intent to disallow it completely. The only variable-ish clause that still allows `augment` is abstract variables, which aren't really variables at all.

> (I'm also not completely sure what `declaration` covers or where it's used. It doesn't seem to include functions.)

The grammar around this seems really messy to me. I *think* all it does is avoid some repeated `;` in clauses but that doesn't seem that useful to me.

I'm really tempted to refactor this entire grammar around members and declarations but I don't know if that will cause any problems because of other specs/tools/tests/whatever depending on specific grammar rules.

  | 'static' 'const' type? initializedIdentifierList ';'
  | 'static' 'final' type? initializedIdentifierList ';'
  | 'static' 'late' 'final' type? initializedIdentifierList ';'
  | 'static' 'late'? varOrType initializedIdentifierList ';'
  | 'covariant' 'late' 'final' type? identifierList ';'
  | 'covariant' 'late'? varOrType initializedIdentifierList ';'
  | 'late'? 'final' type? initializedIdentifierList ';'
  | 'late'? varOrType initializedIdentifierList ';'
  | redirectingFactoryConstructorSignature ';'
  | constantConstructorSignature (redirection | initializers)? ';'

>   | 'augment' constantConstructorSignature initializers? ';'

Author: That was a deliberate choice, yes. By disallowing augmenting redirecting constructors, we avoid all of the complexity around "potentially redirecting" and "potentially non-redirection" which felt really sketchy to me in the proposal.

We could allow augmenting redirecting constructors, but if we do, I'd want some sort of syntax in the introductory declaration that says "this is definitely a redirecting constructor even though the redirection isn't here yet".

>   | 'augment' constantConstructorSignature initializers? ';'

Author: Based on your later comment, I went ahead and added support for filling in an incomplete constructor (generative or factory) with a redirection.

  | constructorSignature (redirection | initializers)? ';'

>   | 'augment' constructorSignature initializers? ';'

Author: I'm going to leave this alone for now (unless the grammar in the proposal is actually incorrect), but I would like to try to do some grammar refactoring in a later change if doing so won't be too disruptive.

```

## Static semantics

### Declaration ordering relations

As part of the meta-programming and augmentation features, we expand the

> [capabilities of part files][parts with imports]. With that feature, a part file

Author: The file is still named parts_with_imports.md. The title in the doc is "Parts with imports" but then it says it defines "enhanced part files". I'll just leave this as-is until I sort that out in a later change. :)

  | 'augment'? 'external' constructorSignature ';'
  | 'augment'? 'external'? 'static'? getterSignature ';'
  | 'augment'? 'external'? 'static'? setterSignature ';'
  | 'augment'? 'external'? 'static'? functionSignature ';'
  | 'external' ('static'? finalVarOrType | 'covariant' varOrType) identifierList ';'
  | 'augment'? 'external'? operatorSignature ';'
  | 'augment'? 'abstract' (finalVarOrType | 'covariant' varOrType) identifierList ';'
  | 'static' 'const' type? initializedIdentifierList ';'
  | 'static' 'final' type? initializedIdentifierList ';'
  | 'static' 'late' 'final' type? initializedIdentifierList ';'
  | 'static' 'late'? varOrType initializedIdentifierList ';'
  | 'covariant' 'late' 'final' type? identifierList ';'
  | 'covariant' 'late'? varOrType initializedIdentifierList ';'
  | 'late'? 'final' type? initializedIdentifierList ';'
  | 'late'? varOrType initializedIdentifierList ';'
  | 'augment'? redirectingFactoryConstructorSignature ';'
  | 'augment'? constantConstructorSignature (redirection | initializers)? ';'
  | 'augment'? constructorSignature (redirection | initializers)? ';'
```

*Note that the grammar for putting `augment` before an extension type

> declaration doesn't allow also specifying a representation field. This is by

Author: Haha, I had to manually diff the grammar in here versus the spec to see what actually changes it had because I didn't notice it either.

        it.*

    *   If the `part` directive in *F* including the file that contains *A* is
        syntactically before the `part` directive in *F* including the file that
        contains *B* in source order, then *A* is before *B*.

    *   Otherwise *B* is before *A*.

Then *B* *is after* *A* if and only if *A* *is before* *B*.

*In short, we complete the partial "is above" order by taking the order of
`part` directives themselves into account when declarations are in sibling
branches of the part tree.*

This order is total (transitive, anti-symmetric, and irreflexive). It
effectively orders declarations by a pre-order depth-first traversal of the file
tree, visiting declarations of a file in source order, and then recursing on
`part` directives in source order.

### Declaration context

Prior to this proposal, an entity like a class or function is introduced by a
single syntactic declaration. With augmentations, an entity may be composed out
of multiple declarations, the introductory one and any number of augmentations.
We define a notion of a *context* to help us talk about the location where we
need to look to collect all of the declarations that define some entity.

> * The context of a top-level declaration is the library and its associated tree

Author: Yeah, I didn't like that term either but it was already in the proposal. I like "augmentation context". (And, having gone through the proposal to make that change, I did find a couple of other unrelated uses of "context", so you're right about it being too general.)

  of part files.

> * The context of a member declaration in a type declaration named *N* is the set

Author: I don't know if "membered" really clarifies much and would then have to be defined. I changed it to "inside a declaration". I think "member" itself should be pretty unambiguous about what it applies to. Let me know if you think that doesn't work.

  of type declarations (introductory and augmenting) named *N* in the enclosing
  set of Dart files.

*Note that context is only defined for the kinds of declarations that can be
augmented. We don't define a context for, say, local variable declarations,
because those aren't subject to augmentation.*

> ### Scoping

Author: I don't think we can make this section fully non-normative because "parts with imports" (correctly) doesn't talk about how identifiers inside type or extension declarations with augmentations are resolved.

This section is honestly kind of a mess. I rewrote it (largely by copying from the original version of this section, ironically).

If a member declaration needs to reference a static or instance member
declared in another introductory or augmenting declaration of the same
type, it can use `this.name` for instance members and `TypeName.name` for
static members to be explicit. Or it can rely on the default if
`name` is not in the lexical scope at all, in which case it's interpreted
as `this.name` if it occurs inside a scope where a `this` is available. _This
approach is always potentially dangerous, since any
third-party import adding a declaration with the same name would break the
code. In practice that's almost never a problem, because instance members

> and top-level declarations usually use different naming strategies._

Reviewer: Didn't we decide at a language meeting that a membered declaration (class, mixin, mixin class, enum, extension type, extension) with the name `N` should have a flat semantic body scope containing all declarations (introductory plus the augmenting) named `N` in the same declaration context?

This would be consistent with the treatment of the top-level scope (also, and perhaps in particular, in the case where the prefix scope is separated from the library scope): It's always a flat namespace, global to the entire Dart file tree.

One thing that falls out is that there is no need to separately explain why we can't declare two unrelated members named `foo` when they are in two different enclosing declarations (we can't rely on the general "two declarations in the same scope with the same name is an error" rule if it isn't one scope).

Author: > Didn't we decide at a language meeting that a membered declaration (class, mixin, mixin class, enum, extension type, extension) with the name `N` should have a flat semantic body scope containing all declarations (introductory plus the augmenting) named `N` in the same declaration context?

You know I honestly don't remember, but digging through the notes, I found a reference to #4080. It seems that, yes, once identifier resolution reaches a membered declaration body, the scope it sees there is the combined scope with all augmentations applied. That's definitely what I want.

Rewrote the "Scoping" section to specify this.

that includes omitting the `extends` keyword. For a variable or parameter,
a `var` keyword may replace the type.

If the type annotation or type parameter bound is omitted in the augmenting
declaration, it is inferred to be the same as the corresponding type annotation
or type parameter bound in the declaration being augmented.

If the type annotation or type parameter bound is *not* omitted, then it's a
**compile-time error** if the type denoted by the augmenting declaration is not

> the same type as the type in the corresponding declaration being augmented.

Author: This is a good question, and definitely a question for @eernstg. :)

I lifted this verbiage straight out of the proposal and moved it here so I think it was written this way deliberately.

words, it's an error to have a declaration marked `augment` with no declaration
to apply it to.*

We say that *I* is the declaration which is *augmented by* *D*.

*In other words, take all of the declarations with the same name in some
context, order them according to after, and each augments the preceding one.
The first one must not be marked `augment` and all the subsequent ones must be.*

An augmentation declaration does not introduce a new name into the surrounding
scope. *We could say that it attaches itself to the existing name.*

### Complete and incomplete declarations

Augmentations aren't allowed to *replace* code, so they mostly add entirely new
declarations to the surrounding type. However, function and constructor
augmentations can fill in a body for an augmented declaration that is lacks one.

More precisely, a function or constructor declaration (introductory or
augmenting) is *incomplete* if all of:

*   The body syntax is `;`.

*   The function is not marked `external`. *An `external` function is considered
    to have a body, just not one that is visible as Dart code.*

*   There is no initializer list. *Obviously, this only applies to constructor

>     declarations.*

Author: Ooh, good catch!

## Applying augmentations

An augmentation declaration *D* is a declaration marked with the built-in
identifier `augment`. We add `augment` as a built-in identifier as a language
versioned change, to avoid breaking pre-feature code.

*D* augments a declaration *I* with the same name and in the same context as
*D*. There may be multiple augmentations in the context of *D*. More precisely,
*I* is the declaration before *D* and after every other declaration before *D*.

It's a **compile-time error** if there is no matching declaration *I*. *In other
words, it's an error to have a declaration marked `augment` with no declaration
to apply it to.*

We say that *I* is the declaration which is *augmented by* *D*.

*In other words, take all of the declarations with the same name in some
context, order them according to the "after" relation, and each augments the
result of all the prior augmentations applied to the original declaration . The
first one must not be marked `augment` and all the subsequent ones must be.*

An augmentation declaration does not introduce a new name into the surrounding
scope. *We could say that it attaches itself to the existing name.*

> ### Complete and incomplete declarations

Reviewer: This simplification feels really, really good!

> Augmentations aren't allowed to *replace* code, so they mostly add entirely new

Reviewer 2: In terms of implementation complexity, it is fairly easy to support this. Actually the ability to add to the extends/with/implements clauses of a class is at a similar if not higher complexity.

Author: That's good to know. I'm still somewhat interested in supporting `augmented()`, but given that we have other more pressing issues, my plan is to leave it out, get this proposal cleaned up and locked down, and move on. We can always revisit augmentations and make them more powerful in the future when we see how users are using them.

declarations to the surrounding type. However, function and constructor
augmentations can fill in a body for an augmented declaration that is lacks one.

More precisely, a function or constructor declaration (introductory or
augmenting) is *incomplete* if all of:

*   The body syntax is `;`.

> *   The function is not marked `external`. *An `external` function is considered

Author: Good point. Filed https://github.com/dart-lang/language/issues/4363 to discuss this further. For now in this PR, I'll leave the behavior as it is which is that `;` is overloaded to mean both abstract and incomplete.

If a declaration is not *incomplete* then it is *complete*.

It's a **compile-time error** if an augmentation is complete and any declaration
before it in the augmentation chain is also complete. *In other words, once a
declaration has acquired a body, no augmentation can replace it with another.*

*It is allowed to augment a complete declaration long as the augmentation itself
is incomplete. This can be useful for an augmentation to add metadata.*

*Examples:*

```dart
a() {}
augment a() {} // Error.

b();
augment b() {} // OK.

c() {}
@meta
augment c(); // OK.

d() {}
augment d(); // OK.
augment d() {} // Error.
```

*Note that the initializer list and body are not treated separately. If a
constructor declaration has an initializer list and `;` body, it is still
considered complete. Likewise, a constructor with no initializer list but a
non-`;` body is complete. Thus a constructor can't acquire an initializer
list in one declaration and a constructor body in another. For example:*

```dart
class C {
  C() : assert(true);
}

augment class C {
  augment C() { body; } // Error. C() is already complete.
}
```

### Application order

> The same declaration can be augmented multiple times by separate augmentation

Reviewer 2: The ability to do this seems somewhat useless if we can't augment complete methods with complete augmentations (and with that invoke the augmented method). In terms of language complexity we seem to spend a look specification work here for no benefit to the user.

Author: It's definitely not *high* benefit. But there are at least two places where augmentation order is user visible:

* Adding mixins (`with` clause) to a class.
* Adding values to an enum.

In most cases, users won't care about or be able to witness augmentation application order, but in a couple of cases they can, so it's gotta be pinned down. Sort of like how instance field initializer order has to be specified because you can see it if the instance fields have initializers with side effects. (Though certainly the specification of that is much simpler than what we have here...)

We could probably say that an augmentation can't add values to an enum unless it has no values (which we also need to add support for #4356). But I don't think we can get away with saying that an augmentation can only add a `with` clause if the type doesn't have one. It would be *really* annoying to not be able to use a mixin on your class just because some code generator happens to be adding an unrelated mixin to your type as well.

}
```

A class, enum, extension type, mixin, or mixin class augmentation may
specify `extends`, `implements` and `with` clauses (when generally

> supported). The types in these clauses are appended to the introductory

Author: Changed the grammar to forbid it.

declarations' clauses of the same kind, and if that clause did not exist
previously, then it is added with the new types.

Instance or static members defined in the body of the augmenting type,
including enum values, are added to the instance or static namespace of the
corresponding type in the introductory declaration. *In other words, the
augmentation can add new members to an existing type.*

Instance and static members inside a class-like declaration may themselves
be augmentations. In that case, they augment the corresponding members in
the same context, according to the rules in the following subsections.

It's a **compile-time** error if:

*   The resulting clauses after being appended would be erroneous if declared

>     directly. This means you can't end up with multiple `extends` clauses on a

Author: > * It's a compile-time error if an augmenting class declaration has an extends clause, and any prior declaration for the same class also has an extends clause.

I like this. Done.

    class, an `on` clause on an enum, etc.

*   A library contains two top-level declarations with the same name, and one of
    the declarations is a class-like declaration and the other is not of the
    same kind, meaning that either one is a class, mixin, enum, extension or
    extension type declaration, and the other is not the same kind of
    declaration.

*   The augmenting declaration and augmented declaration do not have all the
    same modifiers: `abstract`, `base`, `final`, `interface`, `sealed` and

>     `mixin` for `class` declarations, and `base` for `mixin` declarations.

Author: Right. We could theoretically allow it, but it seems very unlikely to be useful and almost certainly confusing for readers.

    *This is not a technical requirement, but follows our design principle that
    what is known from reading the introductory declaration will still be true

>     after augmentation.*

Author: > Adding `mixin` to a `class` with no superclass an no or only trivial constructors, retains all capabilities that were known before.

Fair, but I still think it's a simpler model to just say you keep all the modifiers. We could loosen this somewhat if compelling use cases arise, but it's hard to imagine someone wanting an augmentation to turn a `class` into a `mixin class`.

*   An augmenting extension declares an `on` clause *(this is a syntax

>     error)*. We also do not allow adding further restrictions to a `mixin`

Author: Yeah, this whole error is a little weird. We don't allow an augmenting extension type declaration to contain a representation field, and it's specified by having the grammar directly prohibit it. We don't allow an augmenting mixin declaration to contain an `on` clause and it's *not* specified in the grammar and is instead just made a compile error here. Also, this compile error redundantly points out the thing about extension types.

I'm not sure why these two restrictions were specified differently. I suspect it's because the `on` clause is already optional in a mixin declaration but the representation field is *mandatory* in an introductory extension type declaration and forbidden in an augmenting one.

I went back and specified both in terms of the grammar and then added a note pointing out the syntax change. Then I removed this error because it's now entirely covered by the grammar.

}
```

A class, enum, extension type, mixin, or mixin class augmentation may
specify `extends`, `implements` and `with` clauses (when generally
supported). The types in these clauses are appended to the introductory
declarations' clauses of the same kind, and if that clause did not exist
previously, then it is added with the new types.

*Example:*

```dart
class C implements I1 with M1 {}
augment class C implements I2 with M2 {}

// Is equivalent to:

> class C implements I1, I2 with M1, M2 {}

Reviewer: ```suggestion
class C with M1, M2 implements I1, I2 {}
```

Author: Heh, oops! Changed locally.

*   The augmenting function specifies any default values. *Default values are
    defined solely by the introductory function.*

*   A function is not complete after all augmentations are applied, unless it
    is in a context where it can be abstract. *Every function declaration
    eventually needs to have a body filled in unless it's an instance method
    that can be abstract. In that case, if no declaration provides a body, it
    is considered abstract.*

### Augmenting variables

> A class-like augmentation can add *new* variables (i.e. instance or static

Author: I like "that is".

> fields) to the type being augmented:

Author: I like the formal part being not the parenthetical. The parenthetical is basically explaining "wait, what do you mean by 'variables' inside a class?".

```dart
class C {}

augment class C {
  int x = 3;

  static int y = 4;
}
```

Variable declarations themselves can't be directly augmented, with the exception

> of abstract fields.

Author: Personally, I'm not a fan of conflating an *abstract* declaration (to be filled in by a polymorphic override or implementation) and an *incomplete* declaration (to be filled in by a later augmentation).

The proposal still does that because it seems like `;` is the most natural syntax for an incomplete declaration, but I'd really hesitate to double down on that and use `abstract static` for incomplete static fields that are not abstract at all.

(Though if we go down the path of metaclasses and do something like [`class variables` in Swift](https://www.hackingwithswift.com/example-code/language/whats-the-difference-between-a-static-variable-and-a-class-variable) then maybe `abstract static` starts to make more sense.)

*A variable declaration implicitly has code for the synthesized getter and
setter that access and modify the underlying backing storage. Since we don't
allow augmentations to replace code, that implies that augmentations can't

> change variables. So we don't allow them to be augmented.*

Reviewer 3: That makes sense, except for the lack of ability to provided an initializer expression. But then just make it an abstract variable, and have the augmentation add an actual variable and getter/settter forwarding.

And the `augment abstract ... var` is really just a way to augment with abstract setters and getters.

Author: > But then just make it an abstract variable, and have the augmentation add an actual variable and getter/settter forwarding.

Right. We don't *need* to support augmenting a getter/setter with a variable because you can accomplish the same thing with a little indirection. So I figured I'd keep it simple and get rid of augmenting variables completely.

#### Abstract fields

Dart supports `abstract` field declarations. They are syntactic sugar for
declaring an abstract getter with an optional abstract setter if the variable is
non-final. They don't actually declare a variable with any backing storage.

Because abstract variables are effectively abstract getter and setter
declarations, they can be augmented and used in augmentations just like function
declarations:

*Examples:*

```dart
class C {
  // Augment an abstract variable with a getter:
  abstract final int a;
  augment int get a => 1; // OK.

  // Augment an abstract variable with a getter and setter:
  abstract int b;
  augment int get b => 1; // OK.
  augment set b(int value) {} // OK.

  // Augment a getter with an abstract variable:
  int get c;

  @someMetadata
  augment abstract final int c; // (Not very useful, but valid.)
}

> ```

Reviewer 3: Might need to say how type inheritance work if the getter and setter have different types, and if the declaration has more than one variable:
```dart
int get x;
set x(num _);
abstract String y;
@markOfShame
augment abstract x, y;
```
(Maybe it's not a problem, we use the same inheritance rules as we would for an override in a subclass ... which don't handle this either;
```dart
abstract class C {
  int get x;
  set x(num _);
  abstract String y;
}

abstract class D implements C {
  abstract var x, y;
}
```
makes `x` an error bacause it cannot inherit a combined signature.)

Author: Good catch. Added a section under type annotation inheritance to talk about this.

### Augmenting enum members

An augmentation of an enum type can add new members to the enum, including new
enum values. Enum values are appended in augmentation application order.

Enum values themselves can't be augmented since they are essentially constant

> variables and variables can't be augmented.

Reviewer 3: _(But the static getters can, even if only to add metadata.)_

Author: In this PR, the pitch is that variables can't be augmented *at all*, even by an abstract getter that only adds metadata to the variable. We could allow this, but then it needs to be restricted such that the augmenting getter or setter has to be abstract (incomplete). If you augment a variable with a setter, does the metadata get copied to the variable's implicit getter too?

We could work this out, but I suspect it starts to get special-case-heavy and I don't think it's likely to be very valuable since all you could do is add metadata. So I'm inclined to just leave it out entirely. Variables are *overridable* in the sense of polymorphism and uniform access and all that. But when it comes to augmentations, they're fixed. We could always loosen this in the future if we get a use case.

It's a **compile-time error** if:

*   A declaration inside an augmenting enum declaration has the name `values`,
    `index`, `hashCode`, or `==`. *It has always been an error for an enum
    declaration to declare a member named `index`, `hashCode`, `==`, or
    `values`, and this rule just clarifies that this error is applicable for

>     augmenting declarations as well.*

Reviewer 3: (We should be lifting that rule up to the level of "augmentation chains" or "complete declarations" and say that the *collected* declarations must not contain these members.)

Author: Does the distinction matter? For the collected declarations to contain one of these members, either the introductory declaration declares one (already an error) or one of the augmentations does (an error here).

Also, in general we state that if the result of applying augmentations leads to a combined declaration with errors then those errors are reported. (For example, two augmentations adding members with the same name and conflicting signatures.)

### Augmenting constructors

Augmenting constructors works similar to augmenting a function, with some extra
rules to handle features unique to constructors like initializer lists. Factory
constructors, generative constructors, and const constructors (factory and
generative) can be augmented. Redirecting constructors can't be augmented,
though a class-like augmentation can add *new* redirecting constructors.

*We could support augmenting redirecting constructors, but the use cases seem
rare. Supporting this would require a way to specify that a constructor is
redirecting without actually providing the redirection since the augmentation
needs to fill that in and can't replace an existing redirection in the

> introductory declaration.*

Reviewer 3: A redirecting constructor without a rediction is just, like, a constructor.

I think it should be possible to specify a *signature* of any augmentable declaration, such that the *implementation* can be added by an augmentation.

Whether a generative constructor is initializing or redirecting is an implementation choice.
So, you should be able to declare:
```dart
const Point(int x, int y, {Color? color = Color.black});
```
and have it count as a generative constructor, but not yet a redirecting or initializing one.
Then an augmentation can choose either implementation strategy.

For factory constructors, you should also be able to specify a factory constructor signature:
```dart
factory Point(int x, int y, {Color color = Color.black}); 
```
and then decide whether to implmenet that is forwarding ( `= ...`) or computing (`{ ... }`) in an augmentation.
(Here we run into an issue, because augmentations cannot specify default values, but forwarding constructors also cannot specify default values. That's a problem we made for ourselves, so I believe we can fix it.)

Author: Oh good point. I thought there was something weird about redirections, but I guess not. A redirecting generative constructive is still generative in all the ways that count, so it's harmless to allow completing a generative constructor with a redirection. Likewise for factories.

Done.

*   The augmenting function specifies any default values. *Default values are
    defined solely by the introductory function.*

*   A function is not complete after all augmentations are applied, unless it
    is in a context where it can be abstract. *Every function declaration
    eventually needs to have a body filled in unless it's an instance method
    that can be abstract. In that case, if no declaration provides a body, it
    is considered abstract.*

### Augmenting variables

A class-like augmentation can add *new* variables (that is instance or static
fields) to the type being augmented:

```dart
class C {}

augment class C {
  int x = 3;

  static int y = 4;
}
```

Variable declarations themselves can't be augmented (by variables, getters, or

> setters), with the exception of abstract instance variable declarations.

Reviewer: I guess an abstract `augment` getter could be acceptable, and it could be used to add metadata and DartDoc?

Author: I considered that, but it gets a little weird. If you do:

```dart
int x;

@someMeta
augment int get x;
```

What does the metadata get added *to*? Metadata is a syntactic construct. It doesn't have semantics at all, so there isn't really an "implicit getter" that it makes sense to add it to. Do we add it to the variable itself? Is that even useful?

I couldn't think of a concrete use case for this so I figured I'd just say that variables are entirely closed to augmentation. We can always loosen this in the future if a compelling use case arises, but that seems unlikely.

*A variable declaration implicitly has code for the synthesized getter and
setter that access and modify the underlying backing storage. Since we don't
allow augmentations to replace code, that implies that augmentations can't
change variables. So we don't allow them to be augmented.*

#### Abstract instance variables

> Dart supports `abstract` field declarations. They are syntactic sugar for
> declaring an abstract getter with an optional abstract setter if the variable is
> non-final. They don't actually declare a variable with any backing storage.

Reviewer: The word 'optional' is confusing here. Perhaps:
```suggestion
Dart supports `abstract` field declarations. They are syntactic sugar for
declaring an abstract getter and, if not `final`, an abstract setter.
They don't actually declare a variable with any backing storage.
```

Author: Ooh, much better. Applied locally.

    *Default values are defined solely by the introductory constructor.*

*   The introductory constructor is `const` and the augmenting constructor
    is not or vice versa. *An augmentation can't change whether or not a
    constructor is const because that affects whether users are allowed to use
    the constructor in a const context.*

*   The introductory constructor is marked `factory` and the augmenting
    constructor is not, or vice versa. *An augmentation can't change whether or
    not a constructor is generative because that affects whether users are

>     allowed to call the constructor in a subclass's initializer list.*

Reviewer 3: (That only argues for why *adding* `factory` is bad, not why removing it is.)

Author: An augmentation can't really *remove* anything. If we did allow:

```dart
class C {
  factory C();
}

augment class C {
  augment C() { return ...; }
}
```

Then users would probably reasonably assume that the resulting class took the union of the modifiers like it does with `implements`, `with`, etc.

>     *Default values are defined solely by the introductory constructor.*

Reviewer: I can't comment on line 751, but 'same type,' should probably be 'same type (when provided),'.

Author: Good catch. Done.

other types in which case the extension type syntax will then be understood by
users to be a primary constructor for the extension type.*

[primary constructors]:
https://github.com/dart-lang/language/blob/main/working/2364%20-%20primary%20constructors/feature-specification.md

The extension type's representation object is *not* a variable, even though it
looks and behaves much like one, and it cannot be augmented as such.

It's a **compile-time error** if:

> *   An augmenting declaration has the same name as the representation object.

Reviewer 3: representation object -> representation variable

(The representation object is the value of the representation variable, only the latter has a *name*.)

It's a **compile-time error** if:

It is a compile time error if:
*   An augmenting declaration has the same name as the representation object.

*   An extension type augmentation contains a representation field clause.

[primary constructors]:
https://github.com/dart-lang/language/blob/main/working/2364%20-%20primary%20constructors/feature-specification.md
### Augmenting with metadata annotations

### Augmenting external members

When augmenting an `external` member, it is assumed that a real implementation
of that member has already been filled by some tool prior to any augmentations
being applied. Thus, it is allowed to use `augmented` from augmenting members
on external declarations, but it may throw a `NoSuchMethodError` error at
runtime if no implementation was in fact provided.

**NOTE**: Macros should _not_ be able to statically tell if an external body has
been filled in by a compiler, because it could lead to a different result on
different platforms or tools.

**TODO: Should we add a syntax to let the augmentation dynamically detect
whether there is an external implementation to call?**

> ### Augmenting with metadata annotations and doc comments

Reviewer 4: Why is this removed? This is fairly important

Author: It was redundant. The proposal says elsewhere that this works.

> metadata of the declaration being augmented.

Reviewer 3: I'd say:
> An augmenting declaration can have metadata attached to it.
> Interpretation of metadata is not part of the language,
> and is left to the tools which define the annotations and which
> inspect the source to find them.
> _Some metadata may apply to the combined declaration,
> others to the individual augmenting declarations.
> Only the tool which defined the annotation knows what it means._

Then we can say (somewhere non-normative) that `dart:mirrors` will only work in the combined declarations, and should see the combined annotations, if it works at all.

Author: Good catch. I don't particularly care about mirrors and suspect most users don't either. But I wrote a bit of non-normative stuff about the analyzer package.

### Augmenting with metadata annotations

An augmenting declaration can have metadata attached to it. The language doesn't
specify how metadata is used. Tools may choose to append metadata from
augmentations to the resulting combined declaration or allow inspecting the
metadata on the individual augmentations.

*In practice, most code generators use the [analyzer package][] to introspect
over code. Code generators introspecting on the _syntax_ of some code likely
want to see the metadata for each syntactic declaration separately. Code
generators introspecting over the resolved semantic model of the code (which is
more common) probably want to see the metadata of the introductory declaration
and all augmentations appended into a single list of metadata accessible from
the combined declaration.*

[analyzer package]: https://pub.dev/packages/analyzer

## Dynamic semantics

The application of augmentation declarations to an augmented declaration
produces something that looks and behaves like a single declaration: It has
a single name, a single type or function signature, and it's what all
references to the *name* refers to inside and outside of the library.

> Unlike before, that single *semantic declaration* now consists of multiple

Reviewer: I can't comment on line 896, but I think it would be a good idea to replace non-augmenting by introductory, for consistency.

Reviewer: Line 928 ends in "the *C* doe". That sentence needs to be revised and completed. Also, the occurrence of `abstract` (in code font) in a description of an "abstract method" is confusing: The method is abstract, but it can't contain the keyword `abstract`.

Author: I'll be honest, I'm not even really sure what this "Dynamic Semantics" section is all about. As far as I'm concerned, all we need to do is specified how augmentations are applied and then the resulting declaration's dynamic semantics are exactly as if you had written everything in one declaration (modulo scoping across part files, of course).

This looks like @lrhn's writing style. (@kallentu and I were coincidentally discussing today how we can tell who wrote each part of a spec 😄 ). @lrhn, do you know what this is supposed to be completed with? Do we need this section at all?

_This will usually be exactly the parameter list of the introductory
declaration, but the ordering of named parameters may differ. This is mostly
intended as an example, in practice the augmented parameter list can just be
the parameter list of the introductory declaration, but it's more
direct and clearly correct to use the actual parameter list of the declaration
when creating the parameter scope that its body will run in._

Similarly we define the _augmented function type_ of the declaration stack.
Because of the restrictions we place on augmentations, they will all have the
same function type as the introductory declaration, but again it's

> simpler to assign a function type to every declaration.

Reviewer: Presumably, this step should gather metadata as well?

Author: @lrhn pointed out in his review that metadata doesn't actually have semantics in the language. So I added a section basically saying it's up to tools to decide how metadata is handled but suggesting that it's probably most useful to append it.

> #### Invocation

Reviewer: All references to 'augmented' in this section should be removed, and the rules simplified accordingly.

Author: Done, though again I'm still not really sure what this section is for. :)
